crates/sui-core/src/consensus_adapter.rs

```diff
@@ -240,9 +245,38 @@ impl ConsensusAdapter {
     }
 
     /// Check if this authority should submit the transaction to consensus.
-    fn should_submit(_certificate: &CertifiedTransaction) -> bool {
-        // TODO [issue #1647]: Right now every authority submits the transaction to consensus.
-        true
+    fn should_submit(
+        committee: &Committee,
+        ourselves: &AuthorityName,
+        tx_digest: &TransactionDigest,
+    ) -> bool {
+        // the 32 is as requirement of the deault StdRng::from_seed choice
+        let digest_bytes: [u8; 32] = tx_digest.to_bytes()[..32].try_into().unwrap();
+
+        // permute the validators deterministically, based on the digest
+        let mut rng = StdRng::from_seed(digest_bytes);
+        let mut validators = committee.voting_rights.clone();
+        validators.shuffle(&mut rng);
+
+        // the last (f+1) elements by weight are the submitters for this transaction
+        let mut total_weight = 0u64;
+        let mut submitters = HashSet::new();
+        while total_weight < committee.validity_threshold() {
+            if let Some((name, weight)) = validators.pop() {
+                total_weight += weight;
+                submitters.insert(name);
```

Comment:
Instead of adding the name to submitters, you could directly check if the name is self here.